lib/logstash/outputs/s3.rb

```diff
@@ -439,4 +440,19 @@ def get_under_recovery_files(files)
     end
     skip_files
   end
+
+  # Sorts explored files in desc order
+  # And proceeds bottom-up traversal remove operation if dir is empty
+  # Returns removed set to avoid duplicate Dir.glob process
+  def remove_empty_dirs(files)
+    removed_dirs = Set.new
+    files.sort { |x, y| y.length - x.length } # make sure to remove from leaf node
+         .select { |path| ::File.directory?(path) }
+         .select { |path| (Dir.entries(path) - %w[ . .. ]).empty? } # current and parent dirs escape
+         .each do |path|
+            FileUtils.rm_rf(path, :secure => true)
```

Comment:
If the directory is empty, do we need to use `rm_rf` here, or can we use `rm_f`?

Reply:
The intention was here to avoid handling `StandardError` cases, such as `IOError`.

Reply:
I guess my question was more around the need to use the recursive (and destructive) form of `rm` rather than the `force` functionality. Would you expect that the recursive form is appropriate, and worth the risk of data loss in the event of a coding error?

Reply:
Ah got it. Yup, we don't need recursive operation here. We are starting to remove the leaf of the tree. Our operation itself is recursive, will change it to `FileUtils.rm_f`.